README.md
# _ruski
A scraper tool to download all of God Emporer Putin's interviews on the kremlin website


HISTORY
(base) PS C:\Users\joshh\Documents\Software\_ruski> python .\get_data.py
Meeting with Russian Paralympic team - February 21, 2022 15:00
http://www.en.kremlin.ru/events/president/transcripts/67824 is invalid.
Security Council meeting - February 21, 2022 18:30
http://www.en.kremlin.ru/events/president/transcripts/67826 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67827 is invalid.
Address by the President of the Russian Federation - February 21, 2022 22:35
http://www.en.kremlin.ru/events/president/transcripts/67829 is invalid.
Talks with President of Azerbaijan Ilham Aliyev - February 22, 2022 18:35
http://www.en.kremlin.ru/events/president/transcripts/67831 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67832 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67833 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67834 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67835 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67836 is invalid.
Statements for the press following Russian-Azerbaijani talks - February 22, 2022 19:05
Vladimir Putin answered media questions - February 22, 2022 19:40
Greetings on Defender of the Fatherland Day - February 23, 2022 09:00
http://www.en.kremlin.ru/events/president/transcripts/67840 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67841 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67842 is invalid.
Address by the President of the Russian Federation - February 24, 2022 06:00
http://www.en.kremlin.ru/events/president/transcripts/67844 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67845 is invalid.
Meeting with representatives of Russian business circles - February 24, 2022 19:20
http://www.en.kremlin.ru/events/president/transcripts/67847 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67848 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67849 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67850 is invalid.
Meeting with Security Council permanent members - February 25, 2022 17:45
http://www.en.kremlin.ru/events/president/transcripts/67852 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67853 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67854 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67855 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67856 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67857 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67858 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67859 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67860 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67861 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67862 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67863 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67864 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67865 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67866 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67867 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67868 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67869 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67870 is invalid.
Greetings to military personnel and veterans of Special Operations Forces - February 27, 2022 09:00
http://www.en.kremlin.ru/events/president/transcripts/67872 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67873 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67874 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67875 is invalid.
Meeting with Sergei Shoigu and Valery Gerasimov - February 27, 2022 16:05
The President inspected the progress of construction of a National Space Centre in Moscow - February 27, 2022 18:00
http://www.en.kremlin.ru/events/president/transcripts/67878 is invalid.
Meeting on economic issues - February 28, 2022 15:15
http://www.en.kremlin.ru/events/president/transcripts/67880 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67881 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67882 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67883 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67884 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67885 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67886 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67887 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67888 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67889 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67890 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67891 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67892 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67893 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67894 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67895 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67896 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67897 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67898 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67899 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67900 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67901 is invalid.
http://www.en.kremlin.ru/events/president/transcripts/67902 is invalid.
Meeting with Security Council permanent members - March 3, 2022 19:50

test_downloader.py
import requests
from bs4 import BeautifulSoup


def retrieve_list():
    # Retrieving Kremlin official announcements 
    response = requests.get('http://en.kremlin.ru/events', params={'query': 'interview'}, timeout=40)

    # Checking response status code
    if response.status_code == requests.codes.ok:
        # Output response content as a string
        parse_response(response.text)
    else:
        # Status code is not 200
        print ('Error occured')


def parse_response(html):

    # Initialize a bs4 instance with a HTML-source
    page_dom = BeautifulSoup(html, "html.parser")

    # Iterate over all DIVs with class hentry_event
    for line in page_dom.find_all('div', class_="hentry_event"):
        
        # Find first A inside the DIV and get innerText
        title = line.find('a').contents[0]
        
        # Get article URL
        href = 'http://kremlin.ru'+line.find('a')['href']
        
        # If the line has a TIME node — retrieve 'datetime' attribute
        date = line.find('time')['datetime'] if line.find('time') else 'no-date' 
        
        print (date, title, href)


retrieve_list()

get_data.py
import csv
import requests as req
from bs4 import BeautifulSoup
from time import sleep

def main(MIN, MAX, URL = "http://www.en.kremlin.ru/events/president/transcripts/"):
    """
    Numerical scraping tool to download web urls if they are in numerical order.
     - Takes [MIN | int] to define minimum range.
     - Takes [MAX | int] to define maximum of range.
     - Takes [URL | string] to access string.
    """
    body = ""

    # take all URLS and run on them
    for num in range(MIN, MAX):
        sleep(1)  # if we dont sleep then the kremlin bans us which sucks

        URL = URL + str(num)  # define the url to scrape

        # define the headers for the application to communicate with
        headers = {'User-Agent': 'Mozilla/5.0 \
                (Windows NT 6.1; Win64; x64; rv:47.0) \
                ScrapeBot/3.5)'}

        # make the request
        res = req.get(URL, headers=headers)

        # no point continuing if the page doesn't exist
        if res.status_code != req.codes.ok:
            print(f"{URL} is invalid.")
            continue

        # using bs4 for accessing webpages and downloading content
        soup = BeautifulSoup(res.content, "html.parser")

        try:
            # define variables with try except as they sometimes don't appear in page
            try:
                title = soup.find("h1", class_="entry-title p-name").get_text()
            except AttributeError:
                title = ''
            try:
                date = soup.find("time", class_="read__published").get_text()
            except AttributeError:
                date = ''
            try:
                time = soup.find("div", class_="read__time").get_text()
            except AttributeError:
                time = ''

            # create data dict for csv later
            data = {
                'url':      URL,
                'title':    title,
                'date':     date,
                'time':     time,
                'body':     body
            }

            print(data['title'], '-', data['date'], data['time'])

            for para in soup.find("div", class_="entry-content").find_all("p"):
                try:
                    body += para.get_text()
                except:
                    body = ''

            data['body'] = body

            with open('data.csv', 'a', encoding='utf-8') as df:
                writer = csv.DictWriter(df, fieldnames=data.keys())
                writer.writerow(data)

        except UnicodeEncodeError as err:
            print(err, f"\nAn Error Occurred: at {num}")
            with open("error.txt", 'a', encoding='utf-8') as f:
                f.write(str(num) + '\n')
            continue


if __name__ == "__main__":
    main(67823, 70409)


clean_data.py
import re
import sys
import csv
from time import perf_counter

def clean_body(string):
    """
    Cleans passed text strings for nice readable usage.
    - Takes [filename | string] for the working directory.
    - Takes [TRACE | int] to check whether we are in debug mode.
    """
    print('cleaning...')
    
    string = re.sub("[\xa0\n ]", ' ', string) # remove \n and unicode space
    string = re.sub("[<…>]", '', string)
    string = string.replace('.', '. ').replace('  ', ' ')
    
    return string


def clean_csv(filename='.\data.csv', TRACE=0):
    """
    Cleans CSV files for nice readable formats.
    - Takes [filename | string] for the working directory.
    - Takes [TRACE | int] to check whether we are in debug mode.
    """
    print('cleaning', filename)

    # check performance
    start = perf_counter()
    

    with open(filename, encoding='utf-8') as f_in, \
    open ('clean_data.csv', 'w', newline='', encoding='utf-8') as f_out:
        print(f"File opened: {filename}")

        csv_in = csv.reader(f_in)
        csv_out = csv.writer(f_out)

        n = 0
        for row in csv_in:
            if len(row) > 1:
                row[4] = re.sub("[\xa0\n]", ' ', row[4]) # remove \n and unicode char
                row[4] = re.sub("[<…>]", '', row[4])
                row[4] = row[4].replace('.', '. ')
                csv_out.writerow(row)
            
            x = row[4].split('\n')
            if len(x) > 1:
                print("ERROR: length of string off by too far.")
                sys.exit()
            n+=1
    
    # end timer check
    finished = perf_counter() - start
    
    print(f'Finished cleaning {n} rows in {finished:.6f} seconds')


# if __name__ == "__main__":
#     maxInt = sys.maxsize

#     # decrease maxInt value while OverflowError occurs.
#     while True:
#         try:
#             csv.field_size_limit(maxInt)
#             break
#         except OverflowError:
#             maxInt = int(maxInt/10)
    
#     # clean the csv
#     clean_csv(TRACE=1)

download_urls.py
import requests as req # I prefer terms res and req
from time import sleep


def nume_scrape(URL, MIN, MAX, output="urls.txt"):
    """
    Numerical scraping tool to download web urls if they are in numerical order.
     - Takes [URL | string] to access string.
     - Takes [MIN | int] to define minimum range.
     - Takes [MAX | int] to define maximum of range.
    """

    headers = {'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; \
        x64) AppleWebKit/123.45 (KHTML, like Gecko) \
        Chrome/69.0.6969.69 Safari/537.36'
    }


    for page in range (MIN,MAX):
        sleep(1)
        current_url = URL + '/' + str(page)
        res = req.get(current_url, headers=headers, params={'query': 'interview'}, timeout=40)

        # check page exists and is accessible
        if res.status_code != req.codes.ok:
            continue
        
        # write list to file
        with open(output, 'a', encoding='utf-8') as f:
            f.write(f"\n{current_url}")
        
        print(current_url, res.status_code)

if __name__ == "__main__":
    nume_scrape("http://www.en.kremlin.ru/events/president/transcripts", 65030, 70409)

get_data_to_txt.py
"""
Runs a scraper that downloads from the Kremlin
"""
from json import dumps
import requests
from time import sleep
from bs4 import BeautifulSoup

from clean_data import clean_body

MIN, MAX = 70400, 70410
URL = "http://www.en.kremlin.ru/events/president/transcripts/"


def main():
    """
    Module takes the above url and searches through the defined
    range numerically in order
    """
    print("running")

    # this needs the order [MIN -> MAX] you utter numpty
    for num in range(MIN,MAX):
        sleep(1) # prevents banning

        headers = {'User-Agent': 'Mozilla/5.0 \
                (Windows NT 6.1; Win64; x64; rv:47.0) \
                ScrapeBot/3.5)',
                'Connection': 'keep-alive',
                'Keep-Alive': 'timeout=5,max-100'}

        # make the request
        res = requests.get(f"{URL}{num}", headers=headers)

        # no point continuing if the page doesn't exist
        if res.status_code != requests.codes.okay:
            print(f"{URL}{num} is invalid.")
            continue

        # using bs4 for accessing webpages and downloading content
        soup = BeautifulSoup(res.content, "html.parser")

        try:
            # define variables with try except as they sometimes don't appear in page
            try:
                title = soup.find("h1", class_="entry-title p-name").get_text()
            except AttributeError:
                title = ''
            try:
                date = soup.find("time", class_="read__published").get_text()
            except AttributeError:
                date = ''
            try:
                time = soup.find("div", class_="read__time").get_text()
            except AttributeError:
                time = ''

            # create data dict for csv later
            content = ''
            data = {
                'url':      URL,
                'title':    title,
                'date':     date,
                'time':     time
            }

            print(data['title'], '-', data['date'], '-', data['time'])

            for para in soup.find("div", class_="entry-content").find_all("p"):
                try:
                    content += para.get_text()
                except AttributeError:
                    content = ''

            data['body'] = clean_body(content)

            with open(f"./text_data/{num}_{title}.txt", 'w+', encoding='utf-8') as txt_file:
                txt_file.writelines(data['title']+'\n')
                txt_file.writelines(data['date']+' ')
                txt_file.writelines(data['time']+'\n')
                txt_file.writelines(data['url']+'\n\n')
                txt_file.write(content)

        except UnicodeEncodeError as err:
            print(err, f"\nAn Error Occurred: at {num}")
            with open("error.txt", 'a', encoding='utf-8') as file:
                file.write(str(num) + '\n')
            continue


if __name__ == "__main__":
    main()
